File: concepts/src/main/java/javaexample/chaining/FutureExampleFinished.java
package javaexample.chaining;

import com.twitter.util.Future;
import com.twitter.util.FutureTransformer;

import static javaexample.chaining.FutureDummy.*;

public class FutureExampleFinished {
    public static void main(String[] args) {
        doA().transformedBy(new FutureTransformer<Integer, Integer>() {
            public Future<Integer> flatMap(Integer value) {
                return doB(value);
            }
        }).transformedBy(new FutureTransformer<Integer, Integer>() {
            public Future<Integer> flatMap(Integer value) {
                return doC(value);
            }
        }).transformedBy(new FutureTransformer<Integer, Integer>() {
            public Future<Integer> flatMap(Integer value) {
                return doD(value);
            }

        }).transformedBy(new FutureTransformer<Integer, Void>() {
            public Void map(Integer value) {
                System.out.println(value);
                return null;
            }
        });
    }
}

File: concepts/src/main/scala/scalaexample/chaining/FutureExampleFinished.scala
package scalaexample.chaining

import FutureDummy._

object FutureExampleFinished extends App {
  doA() flatMap doB flatMap doC flatMap doD map {result =>
    println(result)
  }

  // or, with for comprehensions:

  for {
    a <- doA()
    b <- doB(a)
    c <- doC(b)
    d <- doD( c )
  } yield println(d)
}
